File: LeetcodeSwift/LeetcodeSwift/algorithms/q0050.swift
//
//  q0050.swift
//  LeetcodeSwift
//
//  Created by NowOrNever on 02/03/2020.
//  Copyright © 2020 DL. All rights reserved.
//

import Foundation
//50. Pow(x, n)
//Medium
//
//1179
//
//2668
//
//Add to List
//
//Share
//Implement pow(x, n), which calculates x raised to the power n (xn).
//
//Example 1:
//
//Input: 2.00000, 10
//Output: 1024.00000
//Example 2:
//
//Input: 2.10000, 3
//Output: 9.26100
//Example 3:
//
//Input: 2.00000, -2
//Output: 0.25000
//Explanation: 2-2 = 1/22 = 1/4 = 0.25
//Note:
//
//-100.0 < x < 100.0
//n is a 32-bit signed integer, within the range [−231, 231 − 1]
//Accepted
//410,598
//Submissions
//1,411,686
class q0050Solution {
    func myPow(_ x: Double, _ n: Int) -> Double {
        if n == 0 {
            return 1
        }
        var factor = x
        var times = n
        if n < 0 {
            factor = 1.0 / x
            times = -n
        }
        var result = 1.0
        while times > 0 {
            if times & 1 == 1 {
                result *= factor
            }
            times >>= 1
            factor *= factor
        }
        return result
    }
}


func q0050() -> () {
    let solution = q0050Solution()
    let result = solution.myPow(2,10)
    print(result)
}

File: LeetcodePython3/q0050.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if x == 1 or x == 0:
            return x
        if x == - 1:
            if n % 2 == 0:
                return 1
            else:
                return -1
        if n < 0:
            x = 1 / x
            n = -n
        result = 1
        while n != 0:
            if n % 2 != 0:
                result *= x
            n = n // 2
            x = x * x
        return result


solu = Solution()
x = 2.00000
n = -2
result = solu.myPow(x, n)
print(result)